### PR TITLE
Added IMU Task using freeRTOS

### PR DESCRIPTION
**Pull request by harneetMac**

### CODE REVIEW

File: firmware/applications/SignBuddy/main.c [hunk 1]
```diff
@@ -109,22 +109,38 @@ int main(void)
 
   flex_task_setup();
   tsc_task_setup();
+  imu_task_setup();
   sensors_task_setup();
 
   LOG_INFO("App started\r\n");
 
-  RTOS_ERR_CHECK(xTaskCreate(leds_task,
-                             "led_task",
```
Comment (adambujak) [inline firmware/applications/SignBuddy/main.c:117]: why did you remove this?

File: firmware/applications/SignBuddy/main.c [hunk 2]
```diff
@@ -109,22 +109,38 @@ int main(void)
...
+  // }
+  // RTOS_ERR_CHECK(led_task_status);
+
+  // LOG_INFO("Here1\r\n");
 
   sensors_task_start();
+  // LOG_INFO("Here2\r\n");
+
   tsc_task_start();
+  // LOG_INFO("Here3\r\n");
```
Comment (adambujak) [inline firmware/applications/SignBuddy/main.c:134]: can you delete these

File: firmware/applications/SignBuddy/main.c (hunk 1 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/main.c:117]: I didn't really delete it but it's commented out because it was causing an issue. For some reason, LED task is not letting the program to run.

File: firmware/applications/SignBuddy/main.c (hunk 2 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/main.c:134]: yes, i will.

**Comment by stephen99scott** [comment]
I think this was supposed to be a draft PR, there's an issue where the program is crashing when the led task is enabled

File: firmware/applications/SignBuddy/FreeRTOSConfig.h [hunk 3]
```diff
@@ -7,47 +7,47 @@
...
-#define INCLUDE_vTaskDelay                 1
-#define INCLUDE_uxTaskGetStackHighWaterMark 1
+#define INCLUDE_vTaskPrioritySet               1
+#define INCLUDE_uxTaskPriorityGet              1
+#define INCLUDE_vTaskDelete                    1
+#define INCLUDE_vTaskCleanUpResources          1
+#define INCLUDE_vTaskSuspend                   1
+#define INCLUDE_vTaskDelayUntil                1
+#define INCLUDE_vTaskDelay                     1
+#define INCLUDE_uxTaskGetStackHighWaterMark    1
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/FreeRTOSConfig.h:51]: I don't think we want to keep the GetStackHighWaterMark, that was just for debugging

File: firmware/applications/SignBuddy/imu.c [hunk 4]
```diff
@@ -102,23 +101,23 @@ static void imu_task(void *arg)
     sample_data();
     s.callback();
     BaseType_t status = uxTaskGetStackHighWaterMark(NULL);
-    LOG_INFO("status: %d\r\n", (int)status);
+    LOG_INFO("IMU Stack: %d\r\n", (int) status);
   }
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:105]: I think we should delete these too, they're just for debugging

File: firmware/applications/SignBuddy/imu.c [hunk 5]
```diff
@@ -102,23 +101,23 @@ static void imu_task(void *arg)
...
-  // LOG_INFO("Acce dataX: %d\r\n", s.bno055_acce_xyz.x);
-  // LOG_INFO("Acce dataY: %d\r\n", s.bno055_acce_xyz.y);
-  // LOG_INFO("Acce dataZ: %d\r\n", s.bno055_acce_xyz.z);
+  LOG_INFO("Magt calib: %d\r\n", s.mag_calib_stat);
+  LOG_INFO("Accl calib: %d\r\n", s.accel_calib_stat);
+  LOG_INFO("Gyro calib: %d\r\n", s.gyro_calib_stat);
+  LOG_INFO("Quat dataW: %d\r\n", s.bno055_quat_wxyz.w);
+  LOG_INFO("Quat dataX: %d\r\n", s.bno055_quat_wxyz.x);
+  LOG_INFO("Quat dataY: %d\r\n", s.bno055_quat_wxyz.y);
+  LOG_INFO("Quat dataZ: %d\r\n", s.bno055_quat_wxyz.z);
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:117]: Can we return the euler angles instead?

File: firmware/applications/SignBuddy/imu.c [hunk 6]
```diff
@@ -7,8 +7,11 @@
 #include "bno055.h"
 
 #define PROCESS_PERIOD_MS    1000
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:9]: Unusued

File: firmware/applications/SignBuddy/imu.c [hunk 7]
```diff
@@ -7,8 +7,11 @@
...
+#define IMU_SENSOR_CNT       10
 
 typedef struct {
+  TaskHandle_t                   task_handle;
+  void (*callback)(void);
   uint32_t                       last_ticks;
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:15]: Unused

File: firmware/applications/SignBuddy/imu.c [hunk 8]
```diff
@@ -48,6 +51,20 @@ static void hw_init(void)
...
+static inline void sample_data(void)
+{
+  ulTaskNotifyTake(pdFALSE, portMAX_DELAY);
+  uint8_t ret = 0;
+
+  ret |= bno055_get_sys_calib_stat(&s.sys_calib_stat);
+  ret |= bno055_get_mag_calib_stat(&s.mag_calib_stat);
+  ret |= bno055_get_accel_calib_stat(&s.accel_calib_stat);
+  ret |= bno055_get_gyro_calib_stat(&s.gyro_calib_stat);
+  ret |= bno055_read_quaternion_wxyz(&s.bno055_quat_wxyz);
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:63]: Get euler angles

File: firmware/applications/SignBuddy/sensors.c [hunk 9]
```diff
@@ -104,18 +111,19 @@ static void sensors_task(void *arg)
         LOG_DEBUG("Flex_%hu val: %lu\r\n", i + 1, *flex_Ptr);
         flex_Ptr++;
       }
+
+      imu_get_value();
     }
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/sensors.c:116]: Populate imu data into sample structure

File: firmware/applications/SignBuddy/task_config.h [hunk 10]
```diff
@@ -5,10 +5,12 @@
 #define SENSORS_TASK_PRIORITY    (tskIDLE_PRIORITY + 3)
 #define TSC_TASK_PRIORITY        (tskIDLE_PRIORITY + 4)
 #define FLEX_TASK_PRIORITY       (tskIDLE_PRIORITY + 4)
+#define IMU_TASK_PRIORITY        (tskIDLE_PRIORITY + 4)
 
-#define SENSORS_STACK_SIZE       512
+#define SENSORS_STACK_SIZE       256
 #define TSC_STACK_SIZE           256
 #define FLEX_STACK_SIZE          256
+#define IMU_STACK_SIZE           400
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/task_config.h:13]: Reduce back to 256 if that still works

Comment (harneetMac) [reply firmware/applications/SignBuddy/task_config.h:13]: Thanks for the suggestions. I will work on all these and fix them. Most likely tomorrow. I think we may need to keep it at 300 or something because initially it still takes about 260ish.

File: firmware/applications/SignBuddy/imu.c [hunk 11]
```diff
@@ -77,45 +92,58 @@ static void bno_init(void)
...
+  s.imu_values[0] = (int16_t) s.sys_calib_stat;
+  s.imu_values[1] = (int16_t) s.mag_calib_stat;
+  s.imu_values[2] = (int16_t) s.accel_calib_stat;
+  s.imu_values[3] = (int16_t) s.gyro_calib_stat;
+  s.imu_values[4] = s.bno055_euler_hrp.h;
+  s.imu_values[5] = s.bno055_euler_hrp.r;
+  s.imu_values[6] = s.bno055_euler_hrp.p;
+  s.imu_values[7] = s.bno055_acce_xyz.x;
+  s.imu_values[8] = s.bno055_acce_xyz.y;
+  s.imu_values[9] = s.bno055_acce_xyz.z;
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:115]: why are we storing these as an array, when they have meaningful names ?

File: firmware/applications/SignBuddy/sensors.c [hunk 12]
```diff
@@ -106,6 +113,15 @@ static void sensors_task(void *arg)
         LOG_DEBUG("Flex_%hu val: %lu\r\n", i + 1, *flex_Ptr);
         flex_Ptr++;
       }
+
+      uint16_t imu_val;
+      uint32_t *imu_Ptr = (uint32_t *) &s.sample.imuData;
```
Comment (adambujak) [inline firmware/applications/SignBuddy/sensors.c:118]: variable naming case

File: firmware/applications/SignBuddy/sensors.c [hunk 13]
```diff
@@ -78,26 +83,28 @@ static void sensors_task(void *arg)
...
     tsc_start_read();
 
-    //imu_process();
+    imu_start_read();
 
     // TODO Change bits to wait for to all when added
     EventBits_t event_bits;
     event_bits = xEventGroupWaitBits(s.data_ready_event_group,
-                                     TSC_DR_EVENT | FLEX_DR_EVENT,
+                                     TSC_DR_EVENT | FLEX_DR_EVENT | IMU_DR_EVENT,
```
Comment (adambujak) [inline firmware/applications/SignBuddy/sensors.c:91]: use the all sensors event define

File: firmware/applications/SignBuddy/imu.c (hunk 11 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/imu.c:115]: i am not sure how to iterate over a struct and some of them are 8 bits and others are 16 bits. What do you suggest?

File: firmware/applications/SignBuddy/imu.c (hunk 7 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:13]: yeah, unused

Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:13]: last_ticks^

File: firmware/applications/SignBuddy/SignBuddy.proto [hunk 14]
```diff
@@ -9,9 +9,9 @@ message Sample {
     required uint32 sample_id = 1;
 
     message IMUData {
-        required sint32 eul_p = 1;
+        required sint32 eul_h = 1;
         required sint32 eul_r = 2;
-        required sint32 eul_y = 3;
+        required sint32 eul_p = 3;
         required sint32 lin_acc_x = 4;
         required sint32 lin_acc_y = 5;
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/SignBuddy.proto:16]: Need to regenerate proto header and source

File: firmware/applications/SignBuddy/imu.c [hunk 15]
```diff
@@ -77,45 +92,58 @@ static void bno_init(void)
...
-  }
-  s.last_ticks = time;
+void imu_start_read(void)
+{
+  xTaskNotifyGive(s.task_handle);
+}
 
+void imu_get_value(uint16_t *measurement, uint8_t index)
+{
   get_data();
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:141]: unnecessary to get the data to put into an array to put into a different array

File: firmware/applications/SignBuddy/imu.c [hunk 16]
```diff
@@ -48,6 +49,20 @@ static void hw_init(void)
...
 
+static inline void sample_data(void)
+{
+  ulTaskNotifyTake(pdFALSE, portMAX_DELAY);
+  uint8_t ret = 0;
+
+  ret |= bno055_get_sys_calib_stat(&s.sys_calib_stat);
+  ret |= bno055_get_mag_calib_stat(&s.mag_calib_stat);
+  ret |= bno055_get_accel_calib_stat(&s.accel_calib_stat);
+  ret |= bno055_get_gyro_calib_stat(&s.gyro_calib_stat);
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:60]: I don't think we should include the status values in the sample data

**Comment by adambujak** [comment]
rebase this on to master please

File: firmware/applications/SignBuddy/imu.c (hunk 16 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:60]: yeah, agreed

File: firmware/applications/SignBuddy/imu.c (hunk 11 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:115]: just store them with the meaningful names they have
maybe wrapped in a imu_data struct

File: firmware/applications/SignBuddy/SignBuddy.proto (hunk 14 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/SignBuddy.proto:16]: make proto didn't work for me. Am i supposed to do something else before executing it?

File: firmware/applications/SignBuddy/imu.c (hunk 11 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:115]: Yeah, just store them directly in the proto imu data type, and then pass a pointer into get data and copy the data to the pointer

File: firmware/applications/SignBuddy/SignBuddy.proto (hunk 14 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/SignBuddy.proto:16]: You might need to rebuild the docker image if you haven't done that since we added proto, if that doesn't work can you send me the error you are getting?

**Comment by adambujak** [comment]
> rebase this on to master please

i rebased it

**Comment by adambujak** [comment]
this doesn't run

File: firmware/applications/SignBuddy/imu.c (hunk 16 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/imu.c:60]: okay, i will remove these. Do you know of any way to iterate over a struct so i don't have to create a separate array to store values?

Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:60]: Store the values directly in a struct of this type 
image

Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:60]: You can iterate over a struct with a pointer

**Comment by adambujak** [comment]
we've run out of heap space. 

and the timing is completely messed up with the imu added.

**Comment by stephen99scott** [comment]
Can you check the timing of the imu collection by itself? I don't see why it would mess up the timing

File: firmware/applications/SignBuddy/imu.c [hunk 17]
```diff
@@ -48,6 +49,20 @@ static void hw_init(void)
   i2c_init(&s.i2c_instance, IMU_I2C, &i2c_config);
 }
 
+static inline void sample_data(void)
+{
+  ulTaskNotifyTake(pdFALSE, portMAX_DELAY);
+  uint8_t ret = 0;
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:55]: You are calling task notify twice, once here and once on line 98

Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:55]: yeah that'll mess the timing up good

thanks @stephen99scott

File: firmware/applications/SignBuddy/imu.c (hunk 16 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c]: you can look at this code now to see how I would do it

File: firmware/applications/SignBuddy/imu.h [hunk 18]
```diff
@@ -3,6 +3,8 @@
 
 #include <stdint.h>
 
+#include "SignBuddy.pb.h"
+
 #define IMU_SENSOR_CNT    10
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.h:8]: Is IMU_SENSOR_CNT unused?

Comment (adambujak) [reply firmware/applications/SignBuddy/imu.h:8]: yeah idk what that is

**Review by stephen99scott — APPROVED** [review]
Looks good, we should undo that change to the proto file before pushing, or regenerate the headers and rename the variables
